src/dotty/tools/dotc/core/TypeErasure.scala

```diff
@@ -354,22 +356,27 @@ class TypeErasure(isJava: Boolean, isSemi: Boolean, isConstructor: Boolean, wild
         // forwarders to mixin methods.
         // See doc comment for ElimByName for speculation how we could improve this.
       else MethodType(Nil, Nil, eraseResult(rt))
-    case tp => erasure(tp)
+    case tp => this(tp)
+  }
+
+  private def eraseDerivedValueClassRef(tref: TypeRef)(implicit ctx: Context): Type = {
+    val cls = tref.symbol.asClass
+    val underlying = underlyingOfValueClass(cls)
+    ErasedValueType(cls, erasure(underlying, isSemi = false))
   }
 
-  private def eraseDerivedValueClassRef(tref: TypeRef)(implicit ctx: Context): Type =
-    unsupported("eraseDerivedValueClass")
 
   private def eraseNormalClassRef(tref: TypeRef)(implicit ctx: Context): Type = {
     val cls = tref.symbol.asClass
     (if (cls.owner is Package) normalizeClass(cls) else cls).typeRef
   }
 
+  /** The erasure of a function result type. */
   private def eraseResult(tp: Type)(implicit ctx: Context): Type = tp match {
     case tp: TypeRef =>
       val sym = tp.typeSymbol
       if (sym eq defn.UnitClass) sym.typeRef
-      else if (sym.isDerivedValueClass) eraseNormalClassRef(tp)
+      else if (isConstructor && isDerivedValueClass(sym)) eraseNormalClassRef(tp)
```

Comment:
Value classes are unboxed when there is a type mismatch that can be solved by `adaptToType`: if `new Meter(5)` has type `Meter` but the expected type is `ErasedValueType(Meter, Int)` then `adaptToType` will unbox `new Meter(5)`. This is the correct behavior and will not happen if the type of `new Meter(5)` is `ErasedValueType(Meter, Int)` (which would be wrong because the constructor returns an object of type Meter). `clone()` is not a member of `AnyVal` so I don't need to handle it separately, but let me know if you can think of some other case I'd need to handle.